meetings/working-groups/collection-literals/CL-2023-08-03.md

```diff
@@ -0,0 +1,140 @@
+
+```c#
+ReadOnlySpan<int> x1 = [a, b, c]; // Could be local or global scoped
+scoped ReadOnlySpan<int> x2 = [a, b, c]; // Always local scoped.
+```
+
+and then determining the scoping based on how the variable is used.  For example, if 'x1' does not escape, then it could be stack allocated.  If it does end up escaping though (for example with `return x1;`), the collection expression would be heap allocated.  'x2' would always be local scoped of course and would be prevented from escaping.
+
+The benefit of this approach would be allowing the user to always write the simple form with a collection expression, and then get the semantics they needed to make the code legal.  
+
+This was attractive because of the simplicity of the user technically writing simple code, and then not having any stumbling blocks.  However, we felt this was actually a net negative.   A significant problem we forsee with this approach is that users will have code that is fast and nicely stack-allocating as they expect, and then some innocuous looking change ends up causing the compiler to conservatively think it may escape, causing the collection to now be heap allocated.  This could happen silently, easily tanking program performance.  Or, it could come with a warning.  But in the latter case, users would then have to suppress or introduce explicit casts.
+
+Similar to option 1, it felt very strange for us to introduce behavior where the compiler would heap allocate for the user, only to warn since in the majority of times it was believed users would not want that.  After all, in the case of heap allocating, one might as well just choose a heap allocated type *to begin with*.
+
+These concerns led us to our final option:
+
+### Option 3. Always stack-alloc for ref-struct target type.
+
+We then moved to a discussion around our belief that when using ref-structs, and collection expressions, users would expect that data to be put on the stack.  Exploring this led to the following semantics:
+
+```c#
+ReadOnlySpan<int> x1 = [a, b, c];        // stack allocated, local scope.
+scoped ReadonlySpan<int> x2 = [a, b, c]; // stack allocated, local scope.
+```
+
+In other words, the above two have the same meaning.  With the latter just being explicit about the local scoping.  Returning either of those from a method would be illegal.  This has the benefit that the simple form matches the intuition nearly all users will want this to have, and will provide the best performance by default.  This heavily aligns with the stated goal of collection expressions (which we will continue to evangelize it with) that it makes the right default choices and by using it you should always (*ideally*) be getting the best performance results.  
+
+This does have some downsides though. First, the aforementioned redundancy if you actually include `scoped`.  We could consider warning or blocking that to prevent confusion and to help convey that the data is always scoped.
```

Comment:
We have the redundancy today with `scoped ReadOnlySpan<int> x3 = stackalloc int[1];`.

Reply:
will flip.